### DISCUSSION TITLE
Have an npm package name blacklist

### DISCUSSION THREAD

**Discussion by sgpinkus** [npm/feedback#272] [Registry] [answered]
I just caught myself running `npm install help`, instead of the correct `npm help install` and inadvertently installing a package. If the package included a malicious `[pre,post]install]` script, that script would have been run on my system (with default npm settings) if I were not quick enough to Ctrl+C. As a variant on this it's easy to accidentally screw up an install command. Ex `npm i i`, `npm i pack`.

Proposal: NPM should blacklist certain package names that could easily result in a package being installed inadvertently. Seems packages "a" .. "z"  and all the npm sub-command names are taken already, but who knows whats in them now or may be in future. 

Supplemental proposal: More fine grained control over install scripts would help. Setting `ignore-scripts` blocks *every* script from running even if the user explicitly requests it (like `npm run test` etc) so it's a pain keeping this set by default. `ignore-install-scripts: yes|no|prompt` would be nicer. 

----

UPDATE:

Maybe a prompt on install by default setting would also help and be easiest to add. Example, with `apt` Debian package manager, your always prompted unless you provide an `-y` option on command line. Pretty sure you can override in config such that `-y` is always on by default.

**Comment by ljharb** [comment]
A block list wouldn’t fix the root problem - and if a package already exists, then blocking it would break consumers. Certainly if there’s a list of unused but common typos, it would make sense for npm to special-case those on the registry side. This kind of typo list can also be hard to maintain, since they’re not universal errors - everyone is likely to make different typos, especially when coming from different spoken languages.

A “prompt” config setting would be tricky in non-TTY environments.

**Reply by sgpinkus** [reply]
> A “prompt” config setting would be tricky in non-TTY environments.

Just prompt by default. Allow a command line option for non TTY envs to force.

**Comment by MylesBorins** [comment, marked as answer]
@sgpinkus I can definitely see how this could be frustrating, but I don't see how we could explicitly block pacakges that are published to the registry that are not clearly malware, and for packages that are obviously malicious we have a clear and documented process for getting them removed.

We do have a work item already for [auditing packages that overlap with node.js builtins](https://github.com/npm/roadmap/issues/22) which very much aligns with your ask... and we could potentially do a similar work item for all of the existing npm commands.

If you have specific changes to the CLI that you think would be helpful please head over to https://github.com/npm/rfcs to open an issue suggesting a proposal

**Reply by sgpinkus** [reply]
> We do have a work item already for auditing packages that overlap with node.js builtins which very much aligns with your ask... and we could potentially do a similar work item for all of the existing npm commands.

Awesome. I feel like npm commands falls within the scope of this work, and it would be great to handle these in the same work item.

> If you have specific changes to the CLI that you think would be helpful please head over to https://github.com/npm/rfcs to open an issue suggesting a proposal

Yeah for me I think prompt before install if not `--yes` (would probably have to be off by default for backward compat ..), and also an option to disable or prompt on install scripts are actually the best way to deal with it. Will try to make an RFC for this as suggested. Thanks.
